hudi-spark-datasource/hudi-spark3-extensions_2.12/src/main/scala/org/apache/hudi/execution/HudiSQLUtils.scala

```diff
@@ -0,0 +1,553 @@
+        }
+
+        val relationProp = r match {
+          case h: HadoopFsRelation => h.options.updated("path", h.options.getOrElse("hoodie.base.path", ""))
+          case m: MergeOnReadSnapshotRelation => m.optParams.updated("path", m.optParams.getOrElse("hoodie.base.path", ""))
+          case b: HoodieBootstrapRelation => b.optParams.updated("path", b.optParams.getOrElse("hoodie.base.path", ""))
+          case _ => Map.empty[String, String]
+        }
+        relationProp ++ catalogProp
+    }.headOption.getOrElse(Map.empty[String, String])
+  }
+
+  def getRequredFieldsFromTableConf(tablePath: String): (String, Seq[String]) = {
+    val tableConf = tableConfigCache.get(tablePath)
+    val recordKeyFields = tableConf.getProperty(RECORDKEY_FIELD_OPT_KEY).split(",").map(_.trim).filter(!_.isEmpty)
+    val partitionPathFields = tableConf.getProperty(PARTITIONPATH_FIELD_OPT_KEY).split(",").map(_.trim).filter(!_.isEmpty)
+    val preCombineField = tableConf.getProperty(PRECOMBINE_FIELD_OPT_KEY).trim
+    (preCombineField, recordKeyFields ++ partitionPathFields)
+  }
+
+  def merge(
+      df: DataFrame,
+      spark: SparkSession,
+      tableMeta: Map[String, String],
+      trySkipIndex: Boolean = true,
+      enableHive: Boolean = true): Unit = {
+    val savePath = tableMeta.getOrElse("path", throw new HoodieException("cannot find table Path, pls check your hoodie table!!!"))
+    val payload = HudiSQLUtils.tableConfigCache.get(savePath).getProperty(PAYLOAD_CLASS_OPT_KEY, DEFAULT_PAYLOAD_OPT_VAL)
+
+    val df2Merge = payload match {
+      case AWSDMSAVROPAYLOAD =>
+        df.drop("Op").withColumnRenamed(MERGE_MARKER, "Op")
+      case _ =>
+        val f: (String) => Boolean = {x => {if (x == "D") true else false}}
+        val markDeleteUdf = udf(f)
+        df.withColumn(MERGE_MARKER, markDeleteUdf(col(MERGE_MARKER)))
+    }
+    update(df2Merge, buildDefaultParameter(tableMeta, enableHive), spark, WriteOperationType.UPSERT, trySkipIndex, enableHive)
+  }
+
+  def update(
+      df: DataFrame,
+      parameters: Map[String, String],
+      spark: SparkSession,
+      writeOperationType: WriteOperationType = WriteOperationType.UPSERT,
+      trySkipIndex: Boolean = true, enableHive: Boolean = true): Unit = {
+    val tablePath = parameters.get("path").get
+
+    val tableConfig = tableConfigCache.get(tablePath)
+    tableConfig.put(TABLE_TYPE_OPT_KEY, tableConfig.getProperty("hoodie.table.type"))
+    val tblName = tableConfig.get("hoodie.table.name").toString
+    tableConfig.put(HIVE_TABLE_OPT_KEY, tblName)
+    val dataWriteOptions = parameters ++ addSetProperties(parameters, spark) ++ tableConfig.asScala
+
+    checkWriteOptions(dataWriteOptions, enableHive)
+
+    if (df.schema.exists(f => f.name.endsWith(HoodieRecord.FILENAME_METADATA_FIELD)) && trySkipIndex) {
+      log.info("try to upsert table skip index")
+      val par = dataWriteOptions.getOrElse("par", "1500")
```

Comment:
what does `par` mean here?

Reply:
parallelism of hudi operation。   
user has no way to assign parallelism of hudi in sql directly， but i think this parameter can be passed by set statement 。（ set spark.hoodie.shuffle.parallelism=200 ）， so we can  grab parallelism value，and assign it to hudi